### PR TITLE
Windows: use __declspec(thread) TLS implementation, it is MUCH faster…

### PR DESCRIPTION
**Pull request by Araq**
… than _Thread_local [backport]

### CODE REVIEW

File: lib/nimbase.h [hunk 1]
```diff
@@ -125,7 +125,9 @@ __AVR__
   NIM_THREADVAR declaration based on
   http://stackoverflow.com/questions/18298280/how-to-declare-a-variable-as-thread-local-portably
 */
-#if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112 && !defined __STDC_NO_THREADS__
+#if defined _WIN32
+#  define NIM_THREADVAR __declspec(thread)
```
Comment (Araq) [inline lib/nimbase.h]:
```suggestion
#  define NIM_THREADVAR __thread
```

**Comment by RSDuck** [comment]
Why did you change it to use `__thread` on mingw/GCC?

For me `__thread` didn't yield any improvement (with --threads:on more than halving the performance). ~~But then I tried using `__declspec(thread)` and performance is now comparable to Linux, making --threads:on finally useable for me.~~ Apparently this attribute is just ignored with a warning by GCC, so this is not valid.

EDIT 2: apparently gcc does not support native TLS on Windows at all (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=80881), though atleast clang does, so I guess I have to stick with it.

**Comment by Araq** [comment]
I tried clang but even "hello world" crashes with clang on Windows for me. Which clang do you use? I'm willing to change the default C compiler that we ship with Nim if this finally solved this madness.

**Comment by RSDuck** [comment]
I'm using the [one](https://packages.msys2.org/base/mingw-w64-clang) from [msys2](https://msys2.org). While I like its integrated model with its package manager and closed environments, it's probably not for everyone.

One thing of note is that for TLS to work it's mingw clang environment has to be used, as in the default mingw x64 environment all the libraries (and clang) are compiled with gcc, so TLS emulation is used everywhere.

**Pull request merged**